builtins.go

```diff
@@ -839,6 +839,59 @@ func builtinPow(i *interpreter, trace traceElement, basev value, expv value) (va
 	return makeDoubleCheck(i, trace, math.Pow(base.value, exp.value))
 }
 
+func builtinSubstr(i *interpreter, trace traceElement, inputStr, inputFrom, inputLen value) (value, error) {
+	strV, ok := inputStr.(*valueString)
+	if !ok {
+		msg := fmt.Sprintf("substr first parameter should be a string, got %T", inputStr)
```

Comment:
I've just noticed that it will print the Go name for the type in the error message, e.g. "got *jsonnet.valueNumber". We should use Jsonnet names for that: `inputStr.getType().name`. There exist some convenience functions for checking types, such as `i.getNumber`. The errors they produce are not as nice as we have here - they don't include the information which parameter is affected (but perhaps that can be fixed, instead of doing it manually? - probably out of scope for this PR).

Reply:
I'll fix the type part here, but I explicitly used these errors because they match the errors seen in upstream jsonnet stdlib. I thought it would be nice to keep that compatibility.

I'll write some unit tests for this as well.

Reply:
@sbarzowski do you have an example of a unit test for these builtins?

Reply:
It's enough to add an appropriately named `.jsonnet` file to the `testdata/` directory and run `go test -update` to generate output for it (fyi it regenerates output for all the tests).

Reply:
BTW `inputFrom.getType().name` is a string, so you need to use `%s`, not `%T` in the format string in these conditions.

Reply:
> It's enough to add an appropriately named `.jsonnet` file to the `testdata/` directory and run `go test -update` to generate output for it (fyi it regenerates output for all the tests).

Does this only test the "happy" paths? How might I test the "sad" paths?

Reply:
@sbarzowski I see several naming conventions for file names in `testdata` including:

`builtin_<name><#>`
`builtin<Name>_<shortDescription>`
`builtin<Name><ShortDescription>`
`<name>`
`<name>_<short>_<description>`
`std.<name><#>`
`std.<name>_<short>_<description>`

Do you have a preference for file naming here?

Reply:
No strong preference. It's not something I gave much thought so far. 
If they are related to a particular function it's generally it's good to include the verbatim name, with the same capitalization so that it's greppable. I tend to use description instead of numbers these days. 

(The next step for tests organization would be allowing subdirectories and having some high-level categorization. After that we may consider introducing some explicit naming convention for the tests.)